### PR TITLE
Issue/3957 Update reader's software dialog

### PR DESCRIPTION
**Pull request by kidinov**
### Notes

Parent issue #3957 

Design vaw3DvewbUwuQnPXpdNGGH-fi-16%3A673525

~~Merge after https://github.com/woocommerce/woocommerce-android/tree/issue/3957-check-update-available~~

**sorry, GitHub says that it's more than 500 lines, but the PR is pretty straight forward, it shouldn't be that hard to review. If it's needed just let me know and I'll try to split it**

### Description

The PR contains the functionality of the updating software of a reader.

### Success
Just run the app normally
Settings -> Manage readers -> redirect -> connect -> click update in the dialog

https://user-images.githubusercontent.com/4923871/119654996-ba1bef80-be31-11eb-92fd-2d442988458f.mp4

### Fail
Modify code SoftwareUpdateManager::updateSoftware to:

```
    suspend fun updateSoftware() = flow {
        emit(SoftwareUpdateStatus.Initializing)
        checkUpdatesAction.checkUpdates()
        emit(SoftwareUpdateStatus.Failed("updateStatus.e.errorMessage"))
   }
```

https://user-images.githubusercontent.com/4923871/119654809-83de7000-be31-11eb-9f18-b32897775461.mp4

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/detail/CardReaderDetailViewModel.kt [hunk 1]
```diff
@@ -65,6 +70,7 @@ class CardReaderDetailViewModel @Inject constructor(
 
     private fun showConnectedState(readerStatus: Connected, updateAvailable: Boolean = false) {
         viewState.value = if (updateAvailable) {
+            triggerEvent(CardReaderUpdateScreen(skipUpdate = true))
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/detail/CardReaderDetailViewModel.kt]: 🔍 Nitpick: "SkipUpdate" feels a bit misleading to me - we are not skipping the update just letting the screen know that a skip button should be shown, right? What if we renamed it to "startedByUser" or something in that sense? The screen itself can decide what to do.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/detail/CardReaderDetailViewModel.kt [hunk 2]
```diff
@@ -99,13 +105,22 @@ class CardReaderDetailViewModel @Inject constructor(
...
     private fun onUpdateReaderClicked() {
-        // TODO cardreader implement update functionality
+        triggerEvent(CardReaderUpdateScreen(skipUpdate = false))
     }
 
     private fun onDisconnectClicked() {
         // TODO cardreader implement disconnect functionality
     }
 
+    fun onUpdateResult(updateResult: UpdateResult) {
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/detail/CardReaderDetailViewModel.kt]: 🔍 Nitpick: "onUpdateResult" sounds quite generic. I'd consider renaming it to "onUpdateReaderResult" or "onUpdateSoftwareResult" or something similar. Wdyt?

File: WooCommerce/src/main/res/layout/dialog_card_reader_update.xml [hunk 3]
```diff
@@ -0,0 +1,85 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="wrap_content"
+    android:layout_height="wrap_content"
+    android:background="?attr/colorSurface"
+    android:keepScreenOn="true"
```
Comment (malinajirka) [inline WooCommerce/src/main/res/layout/dialog_card_reader_update.xml:8]: Good call, TIL that you can do this using an xml attribute ;).

File: WooCommerce/src/main/res/values-night/colors_base.xml [hunk 4]
```diff
@@ -104,4 +104,10 @@
         Loading Skeletons
     -->
     <color name="skeleton_color">@color/woo_white_alpha_012</color>
+
+    <!--
+        Payment reader
+    -->
+    <color name="update_needed_background_color">@color/woo_orange_70</color>
```
Comment (malinajirka) [inline WooCommerce/src/main/res/values-night/colors_base.xml:111]: We might want to ping garance and ask if we really want to introduce new colors here. It seems `warning_banner_foreground_color` is used for a similar usecase so we might want to keep it consistent. Wdyt?

**Review by malinajirka — APPROVED** [review]
Great job @kidinov! Works as expected and the code looks great ;). I've left a few nitpicky comments, but feel free to ignore them and merge this PR.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/detail/CardReaderDetailViewModel.kt (hunk 2 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/detail/CardReaderDetailViewModel.kt]: Yes, way better. Done https://github.com/woocommerce/woocommerce-android/pull/4082/commits/5168c72fdeb3d1c60bb98c1a4f5fd14ddfb332a5

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/detail/CardReaderDetailViewModel.kt (hunk 1 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/cardreader/detail/CardReaderDetailViewModel.kt]: I had the same thought but probably was tired and didn't fix it. Thanks for noticing. Fixed now =)

File: WooCommerce/src/main/res/values-night/colors_base.xml (hunk 4 above)
Comment (kidinov) [reply WooCommerce/src/main/res/values-night/colors_base.xml:111]: Good point. It's even orange for IOS. I asked [there](p1622461456173000-slack-C6H8C3G23)

Comment (kidinov) [reply WooCommerce/src/main/res/values-night/colors_base.xml:111]: I'll merge it for now, and if the color needs to be changed, I'll change it in another PR. Just not keep this one in a queue =)

**Pull request merged**